src/voter.rs

```diff
@@ -441,16 +480,214 @@ impl<H, N, E: Environment<H, N>> Future for BackgroundRound<H, N, E> where
 
 	fn poll(&mut self) -> Poll<u64, E::Error> {
 		self.task = Some(::futures::task::current());
-		self.inner.poll()?;
 
-		if self.is_done() {
-			Ok(Async::Ready(self.inner.votes.number()))
+		let mut voting_round = self.inner.lock();
+		voting_round.poll()?;
+
+		if self.is_done(&voting_round) {
+			Ok(Async::Ready(voting_round.votes.number()))
 		} else {
 			Ok(Async::NotReady)
 		}
 	}
 }
 
+struct RoundCommitter<H, N, E: Environment<H, N>> where
+	H: Hash + Clone + Eq + Ord + ::std::fmt::Debug,
+	N: Copy + BlockNumberOps + ::std::fmt::Debug,
+{
+	voting_round: Arc<Mutex<VotingRound<H, N, E>>>,
+	commit_timer: E::Timer,
+	last_commit: Option<Commit<H, N, E::Signature, E::Id>>,
+}
+
+impl<H, N, E: Environment<H, N>> RoundCommitter<H, N, E> where
+	H: Hash + Clone + Eq + Ord + ::std::fmt::Debug,
+	N: Copy + BlockNumberOps + ::std::fmt::Debug,
+{
+	fn import_commit(
+		&mut self,
+		env: &E,
+		commit: Commit<H, N, E::Signature, E::Id>,
+	) -> Result<bool, E::Error> {
+		let mut voting_round = self.voting_round.lock();
+
+		// ignore commits for a block lower than we already finalized
+		if commit.target_number < voting_round.votes.finalized().map(|(_, n)| *n).unwrap_or(N::zero()) {
+			return Ok(true);
+		}
+
+		if validate_commit(
+			&commit,
+			voting_round.votes.voters(),
+			voting_round.votes.threshold(),
+			env,
+		)?.is_none() {
+			return Ok(false);
+		}
+
+		// the commit is valid, import all precommits into current round
+		for SignedPrecommit { precommit, signature, id } in commit.precommits.clone() {
+			if let Some(e) = voting_round.votes.import_precommit(env, precommit, id, signature)? {
+				env.precommit_equivocation(voting_round.votes.number(), e);
+			}
+		}
+
+		self.last_commit = Some(commit);
+
+		Ok(true)
+	}
+
+	fn commit(&mut self, env: &E) -> Poll<Option<Commit<H, N, E::Signature, E::Id>>, E::Error> {
+		try_ready!(self.commit_timer.poll());
+
+		let voting_round = self.voting_round.lock();
+		let commit = || -> Option<Commit<H, N, E::Signature, E::Id>> {
+			let (target_hash, target_number) = voting_round.votes.finalized().cloned()?;
+
+			let mut ids = HashSet::new();
+			let precommits =
+				voting_round.votes.precommits().into_iter().filter_map(|(id, precommit, signature)| {
+					match env.ancestry(target_hash.clone(), precommit.target_hash.clone()) {
```

Comment:
this will exclude direct votes on the block I think.
since this is a common pattern we might want to add a new function on `Chain` with provided impl: 
```rust
fn is_equal_or_descendent_of(&self, base: H, block: H) -> Result<bool> {
    if base == block { return Ok(true) }
    
    match self.ancestry(base, block) {
         Ok(_) => Ok(true),
         Err(NotDescendent) => Ok(false),
         Err(e) => Err(e),
    }
}
```

Reply:
I think it works with the current implementation in substrate that uses `tree_route`, but I agree it's better to add that method anyway.